### DISCUSSION TITLE
Modbuss TCP

### DISCUSSION THREAD

**Discussion by RobiMUC** [Lucifer06/RemoteGPIO#41] [Q&A]
I have this relais and this not work 
Is RemoteGPIO not runnint with tcp 
I saw the communication blincking but not work 
what i do wrong

https://www.waveshare.com/wiki/Modbus_POE_ETH_Relay#Overview

**Comment by Lucifer06** [comment]
Did you configure correctly the Dingtian relay module?
Please provide screenshot of both Dingtian relay configuration and RemoteGPIO on Venus OS.

**Reply by RobiMUC** [reply]
![image](https://github.com/user-attachments/assets/6ad64be8-908e-4a84-939a-d0927ef9fd3f)
![image](https://github.com/user-attachments/assets/55d6f60e-113d-47bc-a1aa-6f950be04027)
![image](https://github.com/user-attachments/assets/c0767b20-d4a7-4d1f-bb18-b695f97ee53c)
![image](https://github.com/user-attachments/assets/daee6f48-ad9c-4f44-81fa-dc8553d5e658)

**Comment by RobiMUC** [comment]
Hello i think i found the problem
I use SSCOM tool as a Server on my PC and enable and disable Relais 1 with the [RemoteGPIO]
ON    >>  01 06 00 02 FF 81 A9 9A 
OFF   >>  01 06 00 02 FF 80 68 5A

But the waveshare Relais works with other hex codes
ON    >>  00 00 00 00 00 06 01 05 00 00 FF 00
OFF   >>  00 00 00 00 00 06 01 05 00 00 00 00

Can i change this at a table ??
![image](https://github.com/user-attachments/assets/0c87f40e-35b4-47a2-ad1b-e3f71e7c925a)

**Reply by Lucifer06** [reply]
Hi, RemoteGPIO only supports Dingtian relays. Every vendor is coming with its own protocol (AFAIK there are no standards, protocol wise)....
At some point I may consider adding drivers for other vendors, but it would require additional efforts and standardisation on RemoteGPIO for easy integration with other vendors.

**Reply by RobiMUC** [reply]
maybe a ini file for who i can config the params or a template to build own driver
i have now 3 board and nothing work :-) The description on Amazon don't work but from waveshare
https://amzn.eu/d/8NLp5PC
https://amzn.eu/d/7C9qYHo
https://amzn.eu/d/bSouptn

but i must say the documentation from waveshare is ok
https://www.waveshare.com/wiki/Modbus_RTU_Relay#Function_Code
https://www.waveshare.com/wiki/Modbus_POE_ETH_Relay#Function_Code_Introduction

now i order the same from ali what you describes

**Comment by Schoeppi93** [comment]
Hey guys,
I also ordered the waveshare TCP modbus relais device.
The RemoteGPIO uses the commandline tool "modpoll" for the modbus part so it was kind of straight forward to change the rgpio_driver code to work with it.
The issues here are:
- Modbus has functions for bitwise access or for 16 bit acces. 16 bit access is implemented, so all 8 bits of the relaisboard are set at the same time -> waveshare only supports bit access (Function 5 instead of 6)
- Since the output register is not able to be set at once, you need to iterate over the 8 channels and change the addresses accordingly (I did some bitshifting to do that)

I only did it for the first device on tcp, but maybe that code snipped helps if you are willing to implement this device.
From line 123 I changed the code to:
# Trying to limit resources usage so talking to Unit only if relay change
        if [[ $const != $prev_const1 ]]
                then
                        case $Protocol_unit1 in
                                0) # RS485
                                        /data/RemoteGPIO/bin/modpoll/arm-linux-gnueabihf/modpoll -m rtu -b 115200 -p none -d 8 -1 -r 3 -s 1 -c 1 -a 1 $Port>                                        ;;
                                1) # TCP
                                        i=0
                                        while [ $i -le 8 ]
                                        do
                                                /data/RemoteGPIO/bin/modpoll/arm-linux-gnueabihf/modpoll -m enc -1 -c1 -t0 -r $((i+1)) $IP_Unit1 $(((((const^65280)>>i)&1)) >> /dev/null
                                                i=$((1+i))
                                        done
                                        ;;

Since I usually do stuff like that just for myself one could also do the "only set when changed" stuff for all the bits, but I think the stuff that maybe helps here is the options for modpoll.

So sorry if it could be done in a nicer way, but I hope it helps.

**Comment by Lucifer06** [comment]
I had a look at the Waveshare relay module. It looks to be also a nice device to integrate.
I'm currently finishing RemoteGPIO v4.0 that supports Venus OS 3.6x with Relay and Digital Inputs controlled from Gui v2
RemoteGPIO v4.0 got completely re-written for supporting additional devices and protocols as modules.
Currently I have implemented MQTT, which will be the most efficient in agency and CPU usage.
Waveshare is mentioning MQTT support but can't find any documentation about its capabilities with MQTT...

After I'm done porting the Dingtian-Modbus Module (from RemoteGPIO v3.x), I could consider an interface for a Waveshare-Modbus Module. Let me know if this sounds interesting.

**Reply by Robbie98** [reply]
I bought a waveshare Modbus RTU Relay (B) thinking it would work with this project.... Unfortunately i only just saw this post.... It would be great if you could get it to work!!!

**Reply by Lucifer06** [reply]
If you can wait some time I will have a look.
Meanwhile, you could give a try with an AI assistant, give it both APIs doc (the RemoteGPIO one ant the Waveshare one), ask it to follow the rules and write a driver module. If that works, update the configuration file so we can enable/diable this driver with the virtual switch, and share with the community!

**Reply by Robbie98** [reply]
I only found the remotegpio API.pdf after writing this message. Will give that a test this evening.

**Reply by Robbie98** [reply]
Hi Lucifer, after some difficulties, I've somewhat managed to switch the relays on my Waveshare Modbus RTU relay module. Now it's very unstable and 75% of the time the communication hangs but it's a start. Issue I have at the moment is as soon as I power cycle the Cerbo GX, I loose the /Service/... /Run file. I haven't drilled down far on this yet but if you have any tips. Running REMOTEGPIO 4.0.10 and VenusOS 3.64.

**Reply by Lucifer06** [reply]
Hi Robbie, could you share your driver source code on gitHub? even better, could you add it to the RemotGPIO source code, so this becomes available for all (obviously when we have something working)?
I'll have a look then
